### PR TITLE
Allclose

### LINKED ISSUE #1759: Add allclose to complement isclose
https://numpy.org/doc/stable/reference/generated/numpy.allclose.html#numpy.allclose

can achieve same effect with `sc.all(sc.isclose(...))` Which is actually what is done in `isnear`

### PR DESCRIPTION
**Pull request by OwenArnold**
Fixes #1759

Would be simple to do (even in the python bindings via `isclose` as temporary, however this is implemented as a reduction operation via `apply_in_place`

### CODE REVIEW

File: variable/comparison.cpp [hunk 1]
```diff
@@ -14,12 +15,39 @@ using namespace scipp::core;
 namespace scipp::variable {
 
 namespace {
-Variable _values(Variable &&in) { return in.hasVariances() ? values(in) : in; }
+Variable _values(const Variable &&in) {
```
[the remark this answers is not in the document]
Comment (OwenArnold) [reply variable/comparison.cpp:18]: The const should not be there.

File: variable/comparison.cpp [hunk 2]
```diff
@@ -14,12 +15,39 @@ using namespace scipp::core;
...
+}
+
+Variable tolerance(const Variable &atol, const Variable &rtol,
+                   const Variable &y) {
+  return atol + rtol * abs(y);
+}
 } // namespace
 
+Variable allclose(const Variable &a, const Variable &b, const Variable &rtol,
+                  const Variable &atol, const NanComparisons equal_nans) {
```
[the remark this answers is not in the document]
Comment (OwenArnold) [reply variable/comparison.cpp:29]: I describe that option in the PR description - That was my initial thought, but would be slower and use more memory. I think `allclose` will primarily be used in testing, but It's also possible that users will want it for branching or stop condition for iteration etc, which is why I implemented as a reduction operation.

File: variable/include/scipp/variable/comparison.h [hunk 3]
```diff
@@ -18,4 +18,9 @@ isclose(const Variable &a, const Variable &b, const Variable &rtol,
         const Variable &atol,
         const NanComparisons equal_nans = NanComparisons::NotEqual);
 
+SCIPP_VARIABLE_EXPORT Variable
```
[the remark this answers is not in the document]
Comment (OwenArnold) [reply variable/include/scipp/variable/comparison.h:21]: Yes we don't gain anything from having as a scalar variable I think.

**Comment by OwenArnold** [comment]
I've pushed some clean-up from the review, but maybe this is worth doing this via `sc.all(sc.isclose(...))`? This is more efficient but requires a lot more code..

File: variable/comparison.cpp [hunk 4]
```diff
@@ -15,11 +16,36 @@ namespace scipp::variable {
...
+bool allclose(const Variable &a, const Variable &b, const Variable &rtol,
+              const Variable &atol, const NanComparisons equal_nans) {
+  auto tol = tolerance(atol, rtol, b);
+  if (a.hasVariances() && b.hasVariances()) {
+    return allclose(values(a), values(b), rtol, atol, equal_nans) &
+           allclose(stddevs(a), stddevs(b), rtol, atol, equal_nans);
+  } else {
+    Variable result = makeVariable<bool>(Values{true});
+    if (equal_nans == NanComparisons::Equal)
+      accumulate_in_place(result, const_cast<Variable &>(a), b,
```
Comment (OwenArnold) [inline variable/comparison.cpp:35]: Should figure a way around these `const_casts` too. Maybe just a typo on in the overload for `accumulate_in_place` that can be corrected.

File: variable/comparison.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply variable/comparison.cpp]: - Premature optimization
- More risk for bugs
- Potentially worse for multi-threading, since multi-dim `accumulate_in_place` into single element might not be threaded as well as stepwise, e.g., 2d->1d->0d as performed in `all`.
- More relevant optimization (if it would matter) would be to avoid the *two* copies by `isclose` to `values` and `variances`, instead of the single copy avoided by this, but even that is not worth doing now, I think.

[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply variable/comparison.cpp]: See last item here: https://github.com/scipp/scipp/blob/e873502d5ce3ca6e482bb8a2d93ffc56b913c3af/variable/include/scipp/variable/accumulate.h#L17-L21

That is, the manual "optimization" may actually be a pessimization in this case.

File: python/src/scipp/_comparison.py [hunk 5]
```diff
@@ -148,9 +148,62 @@ def isclose(x, y, rtol=None, atol=None, equal_nan=False):
...
         rtol = 1e-5 * _cpp.units.one
     if atol is None:
         atol = 1e-8 * y.unit
     return _call_cpp_func(_cpp.isclose, x, y, rtol, atol, equal_nan)
+
+
+def allclose(x, y, rtol=None, atol=None, equal_nan=False):
+    """Compares values (x, y) element by element against tolerance absolute
+    and relative tolerances (non-symmetric) then conjoined bitwise across
+    all dimensions (&=) to give a boolean output. The element-wise
```
[the remark this answers is not in the document]
Comment (OwenArnold) [reply python/src/scipp/_comparison.py:165]: Was trying to avoid And'd. I'll reword.

File: variable/comparison.cpp (hunk 4 above)
Comment (OwenArnold) [reply variable/comparison.cpp:35]: Unfortunately it can't: the mutability is used in `bin_detail.h` so we would need a different function.

File: variable/comparison.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (OwenArnold) [reply variable/comparison.cpp]: On my macbook. For 1d inputs ranging 10^0 <= size < 10^6 in length
image
On my macbook. For 2d inputs ranging 10^0 <= size < 10^6 in length (in x, y)
image

File: python/src/scipp/_comparison.py [hunk 6]
```diff
@@ -148,9 +148,60 @@ def isclose(x, y, rtol=None, atol=None, equal_nan=False):
...
+             otherwise False.
+
+    See Also
+    --------
+    :py:func:`scipp.isclose` : comparing element-wise with specified tolerances
+    """
+    if rtol is None:
+        rtol = 1e-5 * _cpp.units.one
+    if atol is None:
+        atol = 1e-8 * y.unit
```
Comment (SimonHeybrock) [inline python/src/scipp/_comparison.py:206]: Why is this needed? Isn't `isclose` handling that?

**Review by SimonHeybrock — APPROVED** [review]
